tests/components/esphome/test_config_flow.py

```diff
@@ -85,156 +90,195 @@ def __init__(self):
+    )
+
+    entry.add_to_hass(hass)
 
-    flow = _setup_flow_handler(hass)
     service_info = {
         "host": "192.168.43.183",
         "port": 6053,
         "hostname": "test8266.local.",
         "properties": {},
     }
-    result = await flow.async_step_zeroconf(user_input=service_info)
-    assert result["type"] == "abort"
+    result = await hass.config_entries.flow.async_init(
+        "esphome", context={"source": "zeroconf"}, data=service_info
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
     assert result["reason"] == "already_configured"
 
+    assert entry.unique_id == "test8266"
+
 
 async def test_discovery_already_configured_ip(hass, mock_client):
     """Test discovery aborts if already configured via static IP."""
-    MockConfigEntry(
-        domain="esphome", data={"host": "192.168.43.183", "port": 6053, "password": ""}
-    ).add_to_hass(hass)
+    entry = MockConfigEntry(
+        domain="esphome",
+        data={CONF_HOST: "192.168.43.183", CONF_PORT: 6053, CONF_PASSWORD: ""},
+    )
+
+    entry.add_to_hass(hass)
 
-    flow = _setup_flow_handler(hass)
     service_info = {
         "host": "192.168.43.183",
         "port": 6053,
         "hostname": "test8266.local.",
         "properties": {"address": "192.168.43.183"},
     }
-    result = await flow.async_step_zeroconf(user_input=service_info)
-    assert result["type"] == "abort"
+    result = await hass.config_entries.flow.async_init(
+        "esphome", context={"source": "zeroconf"}, data=service_info
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
     assert result["reason"] == "already_configured"
 
+    assert entry.unique_id == "test8266"
+
 
 async def test_discovery_already_configured_name(hass, mock_client):
     """Test discovery aborts if already configured via name."""
     entry = MockConfigEntry(
-        domain="esphome", data={"host": "192.168.43.183", "port": 6053, "password": ""}
+        domain="esphome",
+        data={CONF_HOST: "192.168.43.183", CONF_PORT: 6053, CONF_PASSWORD: ""},
     )
     entry.add_to_hass(hass)
+
     mock_entry_data = MagicMock()
     mock_entry_data.device_info.name = "test8266"
     hass.data[DATA_KEY] = {entry.entry_id: mock_entry_data}
 
-    flow = _setup_flow_handler(hass)
     service_info = {
-        "host": "192.168.43.183",
+        "host": "192.168.43.184",
         "port": 6053,
         "hostname": "test8266.local.",
         "properties": {"address": "test8266.local"},
     }
-    result = await flow.async_step_zeroconf(user_input=service_info)
-    assert result["type"] == "abort"
+    result = await hass.config_entries.flow.async_init(
+        "esphome", context={"source": "zeroconf"}, data=service_info
+    )
+
+    assert result["type"] == RESULT_TYPE_ABORT
     assert result["reason"] == "already_configured"
 
+    assert entry.unique_id == "test8266"
+    assert entry.data[CONF_HOST] == "192.168.43.184"
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5451397)